docs(framework) Update `how-to-run-simulations` page (#4492)
adap/flower: doc/source/how-to-run-simulations.rst

===============

Simulating Federated Learning workloads is useful for a multitude of use cases: you
might want to run your workload on a large cohort of clients without having to source,
configure, and manage a large number of physical devices; you might want to run your FL
workloads as fast as possible on the compute systems you have access to without going
through a complex setup process; you might want to validate your algorithm in different
scenarios at varying levels of data and system heterogeneity, client availability,
privacy budgets, etc. These are among some of the use cases where simulating FL
workloads makes sense.

Flower's ``Simulation Engine`` schedules, launches, and manages ``ClientApp`` instances.
It does so through a ``Backend``, which contains several workers (i.e., Python
processes) that can execute a ``ClientApp`` by passing it a ``Context`` and a
``Message``. These ``ClientApp`` objects are identical to those used by Flower's
`Deployment Engine <contributor-explanation-architecture.html>`_, making alternating
between *simulation* and *deployment* an effortless process. The execution of
``ClientApp`` objects through Flower's ``Simulation Engine`` is:

> - **Resource-aware**: Each backend worker executing ``ClientApps`` gets assigned a

Reviewer: ```suggestion
- **Resource-aware**: Each backend worker executing ``ClientApp``\s gets assigned a
```

With the above backend settings, your simulation will run as many ``ClientApps`` in
parallel as CPUs you have in your system. GPU resources for your ``ClientApp`` can be
assigned by specifying the **ratio** of VRAM each should make use of.

.. code-block:: toml

    [tool.flwr.federations.local-simulation]
    options.num-supernodes = 10
    options.backend.client-resources.num-cpus = 1 # each ClientApp assumes to use 1 CPU (default is 2)
    options.backend.client-resources.num-gpus = 0.25 # each ClientApp uses 25% of VRAM (default is 0.0)

.. note::

    If you are using TensorFlow, you need to `enable memory growth
    <https://www.tensorflow.org/guide/gpu#limiting_gpu_memory_growth>`_ so multiple
    ``ClientApp`` instances can share a GPU. This needs to be done before launching the
    simulation. To do so, set the environment variable
    ``TF_FORCE_GPU_ALLOW_GROWTH="1"``.

Let's see how the above configuration results in a different number of ``ClientApps``
running in parallel depending on the resources available in your system. If your system
has:

- 10x CPUs and 1x GPU: at most 4 ``ClientApps`` will run in parallel since each requires
  25% of the available VRAM.

> - 10x CPUs and 2x GPUs: at most 8 ``ClientApps`` will run in parallel.

Reviewer: ```suggestion
- 10x CPUs and 2x GPUs: at most 8 ``ClientApps`` will run in parallel (VRAM-limited).
```

Author: great idea to highlight how they are limited

> - 6x CPUs and 2x GPUs: at most 6 ``ClientApps`` will run in parallel.

Reviewer: ```suggestion
- 6x CPUs and 4x GPUs: at most 6 ``ClientApps`` will run in parallel (CPU-limited).
```

Should we increase this to `4x GPU` to contrast with the previous VRAM-limited example?

- 10x CPUs but 0x GPUs: you won't be able to run the simulation since not even the
  resources for a single ``ClientApp`` can be met.

A generalization of this is given by the following equation. It gives the maximum number
of ``ClientApps`` that can be executed in parallel on available CPU cores (SYS_CPUS) and
VRAM (SYS_GPUS).

.. math::

    N = \min\left(\left\lfloor \frac{\text{SYS_CPUS}}{\text{num_cpus}} \right\rfloor, \left\lfloor \frac{\text{SYS_GPUS}}{\text{num_gpus}} \right\rfloor\right)

> Both ``num_cpus`` (an integer higher than 1) and ``num_gpus`` (a non-negative real

Reviewer 2: ```suggestion
Both ``num_cpus`` (an integer >= 1) and ``num_gpus`` (a non-negative real
```

    # Construct the ClientApp passing the client generation function
    client_app = ClientApp(client_fn=client_fn)

    # Create your ServerApp passing the server generation function
    server_app = ServerApp(server_fn=server_fn)

    run_simulation(
        server_app=server_app,
        client_app=client_app,
        num_supernodes=10,  # equivalent to setting `num-supernodes` in the pyproject.toml
    )

With ``run_simulation``, you can also control the amount of resources for your
``ClientApp`` instances. Do so by setting ``backend_config``. If unset, the default
resources are assigned (i.e., 2xCPUs per ``ClientApp`` and no GPU).

.. code-block:: python

    run_simulation(
        # ...
        backend_config={"client_resources": {"num_cpus": 2, "num_gpus": 0.25}}
    )

Refer to the `30 minutes Federated AI Tutorial
<https://colab.research.google.com/github/adap/flower/blob/main/examples/flower-in-30-minutes/tutorial.ipynb>`_
for a complete example on how to run Flower Simulations in Colab.

Multi-node Flower simulations
-----------------------------

Flower's ``Simulation Engine`` allows you to run FL simulations across multiple compute

> nodes. Before starting your multi-node simulation, ensure that you:

Reviewer: ```suggestion
nodes so that you're not restricted to running simulations on a _single_ machine. Before starting your multi-node simulation, ensure that you:
```

Maybe this is implicit, but it's worth mentioning.

1. Have the same Python environment on all nodes.

> 2. Have a copy of your code (e.g., your entire repo) on all nodes.

Reviewer: ```suggestion
2. Have a copy of your code on all nodes.
```

Probably unnecessary to mention repo here.

3. Have a copy of your dataset on all nodes. If you are using partitions from `Flower

>    Datasets <https://flower.ai/docs/datasets>`_, ensure they are the same.

Reviewer: ```suggestion
   Datasets <https://flower.ai/docs/datasets>`_, ensure the partitioning strategy and configs are the same.
```

Author: 👍  Included, but tweaked it a bit.

4. Start Ray on your head node: on the terminal, type ``ray start --head``. This command
   will print a few lines, one of which indicates how to attach other nodes to the head
   node.

> 5. Attach other nodes to the head node: copy the command shown after starting the head

Reviewer: Should we mention here that all the nodes should be discoverable by the head node?

>    and execute it on the terminal of a new node. For example: ``ray start

Reviewer: ```suggestion
   and execute it on the terminal of a new node (before executing ``flwr run``). For example: ``ray start
```

   --address='192.168.1.132:6379'``.

With all the above done, you can run your code from the head node as you would if the
simulation were running on a single node. In other words:

.. code-block:: shell

    # From your head node, launch the simulation

>     flwr run .

Reviewer: ```suggestion
    flwr run
```

We shouldn't need the `.` now right?

Author: true, i just prefer using it. Removed.

Once your simulation is finished, if you'd like to dismantle your cluster, you simply
need to run the command ``ray stop`` in each node's terminal (including the head node).

.. note::

    When attaching a new node to the head, all its resources (i.e., all CPUs, all GPUs)
    will be visible by the head node. This means that the ``Simulation Engine`` can
    schedule as many ``ClientApp`` instances as that node can possibly run. In some
    settings, you might want to exclude certain resources from the simulation. You can
    do this by appending `--num-cpus=<NUM_CPUS_FROM_NODE>` and/or
    `--num-gpus=<NUM_GPUS_FROM_NODE>` in any ``ray start`` command (including when
    starting the head).

FAQ for Simulations

> -------------------

Reviewer: Let's add a FAQ here about running simulations on multiple machines. This is a question I've seen a few times. We can then link it to your explanation above, where we call it multi-node.

Reviewer: How about also adding an FAQ about tracking results from simulations?

Author: what do you mean by "tracking results" here?

.. dropdown:: Can I make my ``ClientApp`` instances stateful?

    Yes. Use the ``state`` attribute of the `Context <ref-api-flwr.html#flwr.common.Context>`_ object that is passed to the ``ClientApp`` to save variables, parameters, or results to it. Read the `Designing Stateful Clients <how-to-design-stateful-clients.rst>`_ guide for a complete walkthrough.

.. dropdown:: Can I run multiple simulations on the same machine?

    Yes, but bear in mind that each simulation isn't aware of the resource usage of the other. If your simulations make use of GPUs, consider setting the ``CUDA_VISIBLE_DEVICES`` environment variable to make each simulation use a different set of the available GPUs. Export such an environment variable before starting ``flwr run``.

> .. dropdown:: Are the CPU/GPU resources set for each ``ClientApp`` enforced?

Reviewer: ```suggestion
.. dropdown:: Are the CPU/GPU resources set for each ``ClientApp`` enforced?
```

I'm unclear what you mean here by enforce? Do you mean a greedy allocation of resources by the backend?

Author: By "enforce" i mean that even though you may choose `num_gpus=0.25` the simulation engine doesn't restrict the `ClientApp` from using a larger portion of the available VRAM. Earlier in the doc page I use the term _soft_ assignment of resources to convey the same idea. Maybe some other term other than "enforce" should be used?

Author: I've tweaked the question. I think it's clearer now. Thanks for flagging!
